instance/models/appserver.py

```diff
@@ -0,0 +1,254 @@
+# -*- coding: utf-8 -*-
+#
+# OpenCraft -- tools to aid developing and hosting free software projects
+# Copyright (C) 2015 OpenCraft <xavier@opencraft.com>
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+"""
+Instance app models - App Server
+"""
+
+# Imports #####################################################################
+
+import logging
+
+from django.conf import settings
+from django.contrib.contenttypes.models import ContentType
+from django.db import models
+from django.db.models import Q
+from django_extensions.db.models import TimeStampedModel
+
+from instance.logger_adapter import AppServerLoggerAdapter
+from .instance import InstanceReference
+from .log_entry import LogEntry
+from .server import OpenStackServer
+from .utils import ModelResourceStateDescriptor, ResourceState, ValidateModelMixin
+
+
+# Logging #####################################################################
+
+logger = logging.getLogger(__name__)
+
+
+# States ######################################################################
+
+class AppServerState(ResourceState):
+    """
+    A [finite state machine] state describing an instance.
+    """
+    # A state is "steady" if we don't expect it to change.
+    # This information can be used to:
+    # - delay execution of an operation until the target server reaches a steady state
+    # - raise an exception when trying to schedule an operation that depends on a state change
+    #   while the target server is in a steady state
+    # Steady states include:
+    # - Status.New
+    # - Status.Running
+    # - Status.ConfigurationFailed
+    # - Status.Error
+    # - Status.Terminated
+    is_steady_state = True
+
+    # An instance is healthy if it is part of a normal (expected) workflow.
+    # This information can be used to detect problems and highlight them in the UI or notify users.
+    # Healthy states include:
+    # - Status.New
+    # - Status.WaitingForServer
+    # - Status.ConfiguringServer
+    # - Status.Running
+    # - Status.Terminated
+    is_healthy_state = True
+
+
+class Status(ResourceState.Enum):
+    """
+    The states that an instance can be in.
+    """
+
+    class New(AppServerState):
+        """ Newly created """
+        state_id = 'new'
+
+    class WaitingForServer(AppServerState):
```

Comment:
@bradenmacdonald I think it would make sense to rename this to e.g. `WaitingForVM`. Since the class that uses this is now called `AppServer` it is not immediately obvious what `Waiting for server" means (or what "server" refers to). `ConfiguringServer` would have to be renamed as well, e.g. to `ConfiguringVM`.

Thinking about it more, it might be best to try and reserve "server" terminology for app servers, and move to using "VM" terminology for the objects we've been referring to as servers up to now. That would make it easier to distinguish the concepts.

Another approach might be to try and find a different name for `AppServer`, but I haven't been able to come up with a suitable alternative so far.